tests/test_preprocessing/test_match_columns.py

```diff
@@ -126,23 +128,51 @@ def test_drop_columns_when_more_columns_in_test_than_train(df_vartypes, df_na):
     assert match_columns.fill_value is np.nan
     assert match_columns.verbose is True
     assert match_columns.missing_values == "ignore"
+    assert match_columns.match_dtypes is False
     # test fit attrs
     assert list(match_columns.feature_names_in_) == list(train.columns)
     assert match_columns.n_features_in_ == 4
     # test transform output
     pd.testing.assert_frame_equal(expected_result, transformed_df)
 
 
-@pytest.mark.parametrize("fill_value, missing_values, verbose", _params_allowed)
-def test_error_if_param_values_not_allowed(fill_value, missing_values, verbose):
+def test_match_dtypes(df_vartypes):
+    train = df_vartypes.copy()
+    # This creates a test df that contains several incorrect dtypes
+    test = train.astype("category")
+
+    match_columns = MatchVariables(match_dtypes=True)
+    match_columns.fit(train)
+
+    transformed_df = match_columns.transform(test)
+
+    # test init params
+    assert match_columns.match_dtypes is True
+    assert match_columns.fill_value is np.nan
+    assert match_columns.verbose is True
+    assert match_columns.missing_values == "raise"
+    # test fit attrs
+    assert isinstance(match_columns.dtype_dict_, dict)
```

Comment:
I think in this test we can skip all these assert tests, we already have them in previous tests, and focus instead in the functionality being introduced.

In this test, we have 2 datasets, train and test, that are replicas, except that test is then cast as category. This is a relatively safe scenario.

Could we expand this test to cover the following:

- train has datetime as string, test has datetime as datetime
- the reverse of the previous one
- train has numerical variables as string, test as numeric
- the reverse of the above
- train has categorical variables with 3 categories, red green and blue, test has object variable with additional label yellow
- train has object variables with 3 categories, red green and blue, test has categorical variable with additional label yellow

Any test you think we are missing?

Reply:
Should I create all of these tests with the single function "test_match_dtypes"?

I'll remove the asserts.

Reply:
separate tests is usually better, thanks for asking

Reply:
Some of the "reversed" situations change the data so that the data in the dataframes don't match. 

However, I'm thinking that we should NOT test that the data match, only that the dtypes match. I think that I will implement that.

Reply:
so the tests would use this code instead:

pd.testing.assert_series_equal(train.dtypes, transformed_df.dtypes)